packages/nitro-protocol/contracts/StatusManager.sol

```diff
@@ -0,0 +1,89 @@
+pragma solidity 0.7.4;
+
+import './interfaces/IStatusManager.sol';
+
+/**
+ * @dev The StatusManager is responsible for on-chain storage of the status of active channels
+ */
+contract StatusManager is IStatusManager {
+    mapping(bytes32 => bytes32) public statusOf;
+
+    /**
+     * @notice Computes the ChannelMode for a given channelId.
+     * @dev Computes the ChannelMode for a given channelId.
+     * @param channelId Unique identifier for a channel.
+     */
+    function _mode(bytes32 channelId) internal view returns (ChannelMode) {
+        // Note that _unpackStatus(someRandomChannelId) returns (0,0,0), which is
+        // correct when nobody has written to storage yet.
+
+        (, uint48 finalizesAt, ) = _unpackStatus(channelId);
+        if (finalizesAt == 0) {
+            return ChannelMode.Open;
+            // solhint-disable-next-line not-rely-on-time
+        } else if (finalizesAt <= block.timestamp) {
+            return ChannelMode.Finalized;
+        } else {
+            return ChannelMode.Challenge;
+        }
+    }
+
+    /**
+     * @notice Formats the input data for on chain storage.
+     * @dev Formats the input data for on chain storage.
+     * @param channelData ChannelData data.
+     */
+    function _generateStatus(ChannelData memory channelData)
+        internal
+        pure
+        returns (bytes32 status)
+    {
+        // The hash is constructed from left to right.
+        uint256 result;
+        uint16 cursor = 256;
+
+        // Shift turnNumRecord 208 bits left to fill the first 48 bits
+        result = uint256(channelData.turnNumRecord) << (cursor -= 48);
+
+        // logical or with finalizesAt padded with 160 zeros to get the next 48 bits
+        result |= (uint256(channelData.finalizesAt) << (cursor -= 48));
+
+        // logical or with the last 160 bits of the hash the remaining channelData fields
+        // (we call this the fingerprint)
+        result |= uint256(_generateFingerprint(channelData.stateHash, channelData.outcomeHash));
+
+        status = bytes32(result);
+    }
+
+    function _generateFingerprint(bytes32 stateHash, bytes32 outcomeHash)
+        internal
+        pure
+        returns (uint160)
+    {
+        return uint160(uint256(keccak256(abi.encode(stateHash, outcomeHash))));
+    }
```

Comment:
@NiloCK what's the point of this function? Doesn't `stateHash` already contain a commitment to the outcome `outcomeHash`?

Reply:
If we only stored the `stateHash`, we would have to supply the entire preimage of that hash to functions such as `transfer` and `claim`, so that we could decode it to get the outcome. The `state` itself could be much larger than 32 bytes, so use up a lot of calldata and imply a higher gas requirement. 

Having this commitment structure means the gas consumption is independent of the size of the `state`.